### PR TITLE
[CRB-206] Process requests concurrently

### PR DESCRIPTION
**Pull request by nathanwhit**
Currently, only one request can be handled at a time which negatively impacts both throughput and latency. This PR makes it so gateway requests can be handled concurrently. This improves overall performance and makes the gateway's performance more predictable, as slow RPC calls no longer block other requests from being processed.

The performance results (as tested using the procedure described in the load testing documentation) running locally on my machine are:

**Parallel benchmark (with 16 cores):**
Before PR - 8.16 req/sec
With PR   - 105.68 req/sec

**Sequential benchmark:**
Before PR - 8.28 req/sec
With PR   - 8.35 req/sec

In order to make the plugins asynchronous, the plugin interface (defined in `ccplugin`) needed changes. Instead of trying to coordinate two simultaneous PRs, I added a local package (`CCGatewayPlugin`) that defines an asynchronous version of the plugin interface. This can be changed, if desired.

### CODE REVIEW

File: ccgateway/plugins/bitcoin/Bitcoin.cs [hunk 1]
```diff
@@ -73,20 +76,20 @@ public bool Run(IConfiguration cfg, string[] command, out string msg)
                 if (!uint256.TryParse(txId, out var transactionId))
                 {
                     msg = "Invalid transaction: transaction ID invalid";
-                    return false;
+                    return Task.FromResult(Tuple.Create(false, msg));
                 }
 
                 var transactionInfoResponse = rpcClient.GetRawTransactionInfo(transactionId);
```
Comment (seandeneen-gluwa) [inline ccgateway/plugins/bitcoin/Bitcoin.cs]: `await rpcClient.GetRawTransactionInfoAsync` might work better here if we make this into an async method.

**Review by seandeneen-gluwa — CHANGES_REQUESTED** [review]

File: CCGatewayPlugin/CCGatewayPlugin.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+	<TargetFramework>netcoreapp3.1</TargetFramework>
+	<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="5.0.0" />
+  </ItemGroup>
+
+</Project>
```
Comment (atodorov) [inline CCGatewayPlugin/CCGatewayPlugin.csproj:13]: See GitHub Actions log. Docker image can't find this file to build ccgateway.

File: ccgateway/Program.cs [hunk 3]
```diff
@@ -27,20 +27,118 @@
...
+                response = "poor";
+                Console.WriteLine(requestString + ": not enough parameters");
+            }
+            else
+            {
+                string action = command[0];
+                command = command.Skip(1).ToArray();
+
+                ICCGatewayPluginAsync plugin = loader.Get(action);
+                var pluginConfig = config.GetSection(action);
```
Comment (atodorov) [inline ccgateway/Program.cs]: Suggestion: you can move L53 and L56 inside the else block on L61 b/c they aren't needed outside of it (shaving off a few microseconds probably).

**Review by atodorov — CHANGES_REQUESTED** [review]
Docker images fail to build - some changes requested.

**Comment by ghost** [comment]
@nathanwhit LMK when you plan to merge this, because https://dev.azure.com/gluwa/Gluwa/_build?definitionId=128 will need to be updated to match the new folder structure.

**Review by atodorov — APPROVED** [review]

**Review by seandeneen-gluwa — APPROVED** [review]

**Review by CAGS295 — APPROVED** [review]

**Pull request merged**